File: src/0068-Text-Justification/0068.cpp
#include <vector>
#include <string>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    vector<string> fullJustify(vector<string> &words, int L) 
    {
        vector<string> res;
        for(int i = 0, k, l; i < words.size(); i += k) 
        {
            for(k = l = 0; i + k < words.size() and l + words[i+k].size() <= L - k; k++) 
            {
                l += words[i+k].size();
            }
            string tmp = words[i];
            for(int j = 0; j < k - 1; j++) 
            {
                if(i + k >= words.size()) tmp += " ";
                else tmp += string((L - l) / (k - 1) + (j < (L - l) % (k - 1)), ' ');
                tmp += words[i+j+1];
            }
            tmp += string(L - tmp.size(), ' ');
            res.push_back(tmp);
        }
        return res;
    }
};

File: src/0068-Text-Justification/0068.py
class Solution:
    def fullJustify(self, words, maxWidth):
        """
        :type words: List[str]
        :type maxWidth: int
        :rtype: List[str]
        """
        if not words:
            return list()
        
        res, row = list(), list()
        words_len, count = len(words), 0
        i = 0
        while i < words_len:
            if len(words[i]) + count <= maxWidth:
                count += len(words[i]) + 1
                row.append(words[i])
                i += 1
                continue
            else:
                if len(row) == 1:
                    res.append(str(row[0])+' '*(maxWidth - len(row[0])))
                else:
                    extraSpace = maxWidth - count + len(row)
                    num_space = extraSpace//(len(row) - 1)
                    pre_space = extraSpace%(len(row) - 1)
                    tmp = ''
                    for word in row[:-1]:
                        if pre_space > 0:
                            tmp += str(word) + ' '*(num_space + 1)
                            pre_space -= 1
                        else:
                            tmp += str(word) + ' '*(num_space + 0)
                            
                    tmp += str(row[-1])
                    res.append(tmp)
                    
                row.clear()
                count = 0
                
        tmp = ' '.join(row)  
        res.append(tmp + (maxWidth - len(tmp))*' ')
        return res

if __name__ == "__main__":
    words = ["What","must","be","acknowledgment","shall","be"]
    maxWidth = 16
    print(Solution().fullJustify(words, maxWidth))